### PR TITLE
3.x: [Java 8] Upgrade to Java 8, add Flowable.fromX operators

### PR DESCRIPTION
**Pull request by akarnokd**
This PR upgrades RxJava to use and compile with Java 8.

- Build target set to 8
- AnimalSniffer set to 8
- JavaDocs link to the official JDK set to 8
- Implemented operators:
  - `Flowable.fromOptional`
  - `Flowable.fromCompletionStage`
  - `Flowable.fromStream`
- Applied `@SafeVarargs` to many methods
- Created marbles

Related #6695

image
image
image

### CODE REVIEW

File: src/main/java/io/reactivex/rxjava3/core/Flowable.java [hunk 1]
```diff
@@ -18620,4 +18619,121 @@ public final Completable switchMapCompletableDelayError(@NonNull Function<? supe
...
+     * @param optional the optional value to convert into a {@code Flowable}
+     * @return the new Flowable instance
+     * @see #just(Object)
+     * @see #empty()
+     * @since 3.0.0
+     */
+    @CheckReturnValue
+    @BackpressureSupport(BackpressureKind.FULL)
+    @SchedulerSupport(SchedulerSupport.NONE)
+    public static <T> Flowable<T> fromOptional(@NonNull Optional<T> optional) {
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/rxjava3/core/Flowable.java:18654]: +NonNull for return type

File: src/main/java/io/reactivex/rxjava3/core/Flowable.java [hunk 2]
```diff
@@ -18620,4 +18619,121 @@ public final Completable switchMapCompletableDelayError(@NonNull Function<? supe
...
+     * </dl>
+     * @param <T> the element type of the CompletionStage
+     * @param stage the CompletionStage to convert to Flowable and signal its terminal value or error
+     * @return the new Flowable instance
+     * @since 3.0.0
+     */
+    @CheckReturnValue
+    @BackpressureSupport(BackpressureKind.FULL)
+    @SchedulerSupport(SchedulerSupport.NONE)
+    public static <T> Flowable<T> fromCompletionStage(@NonNull CompletionStage<T> stage) {
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/rxjava3/core/Flowable.java:18690]: +NonNull for return type

File: src/main/java/io/reactivex/rxjava3/core/Flowable.java [hunk 3]
```diff
@@ -18620,4 +18619,121 @@ public final Completable switchMapCompletableDelayError(@NonNull Function<? supe
...
+     * @param <T> the element type of the source {@code Stream}
+     * @param stream the {@code Stream} of values to emit
+     * @return the new Flowable instance
+     * @since 3.0.0
+     * @see #fromIterable(Iterable)
+     */
+    @CheckReturnValue
+    @BackpressureSupport(BackpressureKind.FULL)
+    @SchedulerSupport(SchedulerSupport.NONE)
+    public static <T> Flowable<T> fromStream(@NonNull Stream<T> stream) {
```
Comment (JakeWharton) [inline src/main/java/io/reactivex/rxjava3/core/Flowable.java:18735]: +NonNull for return type

**Review by JakeWharton — APPROVED** [review]

File: src/main/java/io/reactivex/rxjava3/core/Flowable.java (hunk 1 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/rxjava3/core/Flowable.java:18654]: Done. I suppose it should be added to all other methods too.

Comment (vanniktech) [reply src/main/java/io/reactivex/rxjava3/core/Flowable.java:18654]: That plus I think Jake meant `<T> Flowable<@NonNull T>`

Comment (akarnokd) [reply src/main/java/io/reactivex/rxjava3/core/Flowable.java:18654]: Okay, the other methods will get these in some other PRs.

File: src/main/java/io/reactivex/rxjava3/annotations/NonNull.java [hunk 4]
```diff
@@ -19,10 +19,10 @@
 import java.lang.annotation.*;
 
 /**
- * Indicates that a field/parameter/variable/return type is never null.
```
Comment (vanniktech) [inline src/main/java/io/reactivex/rxjava3/annotations/NonNull.java:22]: Do we want to mention type use too?

**Review by vanniktech — APPROVED** [review]

File: src/main/java/io/reactivex/rxjava3/annotations/NonNull.java (hunk 4 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/rxjava3/annotations/NonNull.java:22]: It's not that important; type parameter should imply its use too.

**Comment by He-Pin** [comment]
With this move, there are many we can help.

**Comment by He-Pin** [comment]
Should remove the usage of `Optional.of` with `Optional.ofNullable`.

**Pull request merged**